File: bucky.m
% Code Matlab mo phong he dong thuan voi do thi G1
A=zeros(60,60);
 for i=1:59
  A(i,i+1)=1;
  A(i+1,i)=1;
 end
A(60,1)=1;
A(1,60)=1;
G = graph(A);
 global L
 L=laplacian(G)*eye(60);

 % Giai phuong trinh vi phan dung ode45
x0 = 10*(rand(60,1)-0.5); % Dieu kien dau
 [t,x] = ode45(@control_law,[0 100],x0);

 % Bieu dien tren do thi
 figure(2);hold on;
 for i=1:60
 plot(t,x(:,i)','LineWidth',1);
 end
xlabel 'Thoi gian [s]';
ylabel 'Bien trang thai x_i(t)';
 title 'x_i(t)'
 box on;

%% Ham tinh luat dong thuan
 function dpdt = control_law(t,p)
 global L
 dpdt = -L*p;
 end

File: bucky.py
import matplotlib.pyplot as plt
from scipy.sparse import csgraph
from scipy.integrate import odeint
import numpy as np
from numpy import array, dot, pi
import random
def deriv(x,t):            
    return dot(-L,x)
s = (60,60)
A = np.zeros(s)
for i in range(59):
    A[i][i+1]=1
    A[i+1][i]=1
A[0][59]=1
A[59][0]=1
print(A)
L = csgraph.laplacian(A, normed=False)
t = np.linspace(0, 100, 200)
s = (1, 60)
x0 = np.zeros(s)

for i in range(60):
    x0[0][i]=random.uniform(-5.0, 5.0)
x0 = x0.reshape(-1);
MA = odeint(deriv, x0, t)
for i in range(60):
    plt.plot(t,MA[:,i])
plt.title('Bucky')
plt.xlabel('Thoi gian[x]')
plt.ylabel('Bien trang thai x_i(t)')
plt.show()

